Lib9c/Delegation/RewardBase.cs

```diff
@@ -0,0 +1,395 @@
+        /// <param name="share">
+        /// The number of shares to calculate the reward.
+        /// </param>
+        /// <param name="currency">
+        /// The currency to calculate the reward.
+        /// </param>
+        /// <returns>
+        /// Cumulative reward during the period, for the specific currency.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the <paramref name="currency"/> is not in the <see cref="RewardPortion"/>.
+        /// </exception>
+        public FungibleAssetValue CumulativeRewardDuringPeriod(BigInteger share, Currency currency)
+            => RewardPortion.TryGetValue(currency, out var portion)
+                ? (portion * share).DivRem(Multiplier(SigFig)).Quotient
+                : throw new ArgumentException($"Invalid reward currency: {currency}");
+
+        private static RewardBase AddReward(RewardBase rewardBase, FungibleAssetValue reward)
+            => new RewardBase(
+                rewardBase.Address,
+                rewardBase.TotalShares,
+                rewardBase.RewardPortion.TryGetValue(reward.Currency, out var portion)
+                    ? rewardBase.RewardPortion.SetItem(
+                        reward.Currency,
+                        portion + (reward * Multiplier(rewardBase.SigFig)).DivRem(rewardBase.TotalShares).Quotient)
+                    : throw new ArgumentException($"Invalid reward currency: {reward.Currency}"),
+                rewardBase.SigFig,
+                rewardBase.StartHeight);
+
+        private static RewardBase UpdateTotalShares(RewardBase rewardBase, BigInteger totalShares)
+        {
+            var newSigFig = Math.Max(rewardBase.SigFig, RecommendedSigFig(totalShares));
+            var multiplier = Multiplier(newSigFig - rewardBase.SigFig);
+            var newPortion = rewardBase.RewardPortion.ToImmutableSortedDictionary(
+                kvp => kvp.Key,
+                kvp => kvp.Value * multiplier,
+                _currencyComparer);
+
+            return new RewardBase(
+                rewardBase.Address,
+                totalShares,
+                newPortion,
+                newSigFig);
+        }
+
+        private static int RecommendedSigFig(BigInteger totalShares)
+            => (int)Math.Floor(BigInteger.Log10(totalShares)) + Margin;
+
+        private static BigInteger Multiplier(int sigFig)
+            => BigInteger.Pow(10, sigFig);
+
+
+        public override bool Equals(object? obj)
+            => obj is RewardBase other && Equals(other);
+
+        public bool Equals(RewardBase? other)
+            => ReferenceEquals(this, other)
+            || (other is RewardBase rewardBase
+            && Address == rewardBase.Address
+            && TotalShares == rewardBase.TotalShares
+            && RewardPortion.SequenceEqual(rewardBase.RewardPortion)
+            && SigFig == rewardBase.SigFig);
+
+        public override int GetHashCode()
+            => Address.GetHashCode();
```

Comment:
I think the code below is correct, is there anything wrong with it?

```suggestion
            => HashCode.Combine(Address, TotalShares, RewardPortion, SigFig, StartHeight);
```

Reply:
I thought `Address` is enough to represent its identity.